packages/form-js-viewer/src/render/hooks/useValuesAsync.js

```diff
@@ -35,11 +33,8 @@ export default function(field) {
   const [ valuesGetter, setValuesGetter ] = useState({ values: [], error: undefined, state: LOAD_STATES.LOADING });
   const initialData = useService('form')._getState().initialData;
 
-  const evaluatedValues = valuesExpression ? (useExpressionEvaluation(valuesExpression) || []) : [];
-
-  // re-calculate values if evaluated expression changes
-  const previousEvaluatedValues = usePrevious(evaluatedValues, [], [ evaluatedValues ]);
-  const evaluatedValuesChanged = !compare(previousEvaluatedValues, evaluatedValues);
+  const expressionEvaluation = useExpressionEvaluation(valuesExpression);
+  const evaluatedValues = useDeepCompareState(expressionEvaluation || [], []);
```

Comment:
I abstracted away all that funky compare logic into its own hook, makes everything cleaner. Plus we may want to re-use this deep comparison stuff other places, for example here:

![grafik](https://github.com/bpmn-io/form-js/assets/17801113/0e35fe7c-017a-4d15-adf9-d6fcf6c9e393)

Reply:
Thanks a lot 🏅